File: regions/system/library/typescript/migration/__tests__/diff.test.ts
import { describe, it, expect } from 'vitest';
import { diffSchemas } from '../src/diff.js';
import type { Schema, Column, Table } from '../src/schema.js';

function makeColumn(
  name: string,
  dataType: string,
  nullable: boolean,
): Column {
  return { name, dataType, nullable };
}

function makeTable(name: string, columns: Column[]): Table {
  return { name, columns, indexes: [], constraints: [] };
}

describe('diffSchemas', () => {
  it('should detect table added', () => {
    const oldSchema: Schema = { tables: [] };
    const newSchema: Schema = {
      tables: [makeTable('users', [makeColumn('id', 'UUID', false)])],
    };
    const diffs = diffSchemas(oldSchema, newSchema);
    expect(diffs).toHaveLength(1);
    expect(diffs[0].type).toBe('tableAdded');
    if (diffs[0].type === 'tableAdded') {
      expect(diffs[0].table.name).toBe('users');
    }
  });

  it('should detect table dropped', () => {
    const oldSchema: Schema = {
      tables: [makeTable('users', [makeColumn('id', 'UUID', false)])],
    };
    const newSchema: Schema = { tables: [] };
    const diffs = diffSchemas(oldSchema, newSchema);
    expect(diffs).toHaveLength(1);
    expect(diffs[0].type).toBe('tableDropped');
    if (diffs[0].type === 'tableDropped') {
      expect(diffs[0].tableName).toBe('users');
    }
  });

  it('should detect column added', () => {
    const oldSchema: Schema = {
      tables: [makeTable('users', [makeColumn('id', 'UUID', false)])],
    };
    const newSchema: Schema = {
      tables: [
        makeTable('users', [
          makeColumn('id', 'UUID', false),
          makeColumn('email', 'TEXT', true),
        ]),
      ],
    };
    const diffs = diffSchemas(oldSchema, newSchema);
    expect(diffs).toHaveLength(1);
    expect(diffs[0].type).toBe('columnAdded');
    if (diffs[0].type === 'columnAdded') {
      expect(diffs[0].table).toBe('users');
      expect(diffs[0].column.name).toBe('email');
    }
  });

  it('should detect column dropped', () => {
    const oldSchema: Schema = {
      tables: [
        makeTable('users', [
          makeColumn('id', 'UUID', false),
          makeColumn('email', 'TEXT', true),
        ]),
      ],
    };
    const newSchema: Schema = {
      tables: [makeTable('users', [makeColumn('id', 'UUID', false)])],
    };
    const diffs = diffSchemas(oldSchema, newSchema);
    expect(diffs).toHaveLength(1);
    expect(diffs[0].type).toBe('columnDropped');
    if (diffs[0].type === 'columnDropped') {
      expect(diffs[0].table).toBe('users');
      expect(diffs[0].columnName).toBe('email');
    }
  });

  it('should detect column changed', () => {
    const oldSchema: Schema = {
      tables: [makeTable('users', [makeColumn('name', 'TEXT', true)])],
    };
    const newSchema: Schema = {
      tables: [makeTable('users', [makeColumn('name', 'VARCHAR', false)])],
    };
    const diffs = diffSchemas(oldSchema, newSchema);
    expect(diffs).toHaveLength(1);
    expect(diffs[0].type).toBe('columnChanged');
    if (diffs[0].type === 'columnChanged') {
      expect(diffs[0].table).toBe('users');
      expect(diffs[0].columnName).toBe('name');
      expect(diffs[0].from.dataType).toBe('TEXT');
      expect(diffs[0].to.dataType).toBe('VARCHAR');
    }
  });

  it('should return empty array when no changes', () => {
    const schema: Schema = {
      tables: [makeTable('users', [makeColumn('id', 'UUID', false)])],
    };
    const diffs = diffSchemas(schema, schema);
    expect(diffs).toHaveLength(0);
  });
});

File: regions/system/library/dart/migration/test/evolution/diff_test.dart
import 'package:test/test.dart';
import 'package:k1s0_migration/migration.dart';

Column makeColumn(String name, String dataType, bool nullable) {
  return Column(name: name, dataType: dataType, nullable: nullable);
}

Table makeTable(String name, List<Column> columns) {
  return Table(name: name, columns: columns);
}

void main() {
  group('diffSchemas', () {
    test('テーブルの追加が検出されること', () {
      final oldSchema = Schema(tables: []);
      final newSchema = Schema(tables: [
        makeTable('users', [makeColumn('id', 'UUID', false)]),
      ]);
      final diffs = diffSchemas(oldSchema, newSchema);
      expect(diffs.length, 1);
      expect(diffs[0], isA<TableAddedDiff>());
      expect((diffs[0] as TableAddedDiff).table.name, 'users');
    });

    test('テーブルの削除が検出されること', () {
      final oldSchema = Schema(tables: [
        makeTable('users', [makeColumn('id', 'UUID', false)]),
      ]);
      final newSchema = Schema(tables: []);
      final diffs = diffSchemas(oldSchema, newSchema);
      expect(diffs.length, 1);
      expect(diffs[0], isA<TableDroppedDiff>());
      expect((diffs[0] as TableDroppedDiff).tableName, 'users');
    });

    test('カラムの追加が検出されること', () {
      final oldSchema = Schema(tables: [
        makeTable('users', [makeColumn('id', 'UUID', false)]),
      ]);
      final newSchema = Schema(tables: [
        makeTable('users', [
          makeColumn('id', 'UUID', false),
          makeColumn('email', 'TEXT', true),
        ]),
      ]);
      final diffs = diffSchemas(oldSchema, newSchema);
      expect(diffs.length, 1);
      expect(diffs[0], isA<ColumnAddedDiff>());
      final diff = diffs[0] as ColumnAddedDiff;
      expect(diff.table, 'users');
      expect(diff.column.name, 'email');
    });

    test('カラムの削除が検出されること', () {
      final oldSchema = Schema(tables: [
        makeTable('users', [
          makeColumn('id', 'UUID', false),
          makeColumn('email', 'TEXT', true),
        ]),
      ]);
      final newSchema = Schema(tables: [
        makeTable('users', [makeColumn('id', 'UUID', false)]),
      ]);
      final diffs = diffSchemas(oldSchema, newSchema);
      expect(diffs.length, 1);
      expect(diffs[0], isA<ColumnDroppedDiff>());
      final diff = diffs[0] as ColumnDroppedDiff;
      expect(diff.table, 'users');
      expect(diff.columnName, 'email');
    });

    test('カラムの変更が検出されること', () {
      final oldSchema = Schema(tables: [
        makeTable('users', [makeColumn('name', 'TEXT', true)]),
      ]);
      final newSchema = Schema(tables: [
        makeTable('users', [makeColumn('name', 'VARCHAR', false)]),
      ]);
      final diffs = diffSchemas(oldSchema, newSchema);
      expect(diffs.length, 1);
      expect(diffs[0], isA<ColumnChangedDiff>());
      final diff = diffs[0] as ColumnChangedDiff;
      expect(diff.table, 'users');
      expect(diff.columnName, 'name');
      expect(diff.from.dataType, 'TEXT');
      expect(diff.to.dataType, 'VARCHAR');
    });

    test('同一スキーマの場合に差分がないこと', () {
      final schema = Schema(tables: [
        makeTable('users', [makeColumn('id', 'UUID', false)]),
      ]);
      final diffs = diffSchemas(schema, schema);
      expect(diffs, isEmpty);
    });
  });
}
